Refactored the overworld scanning discord bot and

jsb-pokemon opened the pull request:
added more detailed notifications.

* The discord bot is its own class
* The test is now named inline with other tests
* The test is a minimal stub to invoke the discord bot
* The config file now takes Channel IDs for both text and event based notifications.

jsb-pokemon committed:
Refactored the overworld scanning discord bot and
added more detailed notifications.

* The discord bot is its own class
* The test is now named inline with other tests
* The test is a minimal stub to invoke the discord bot
* The config file now takes Channel IDs for both text and event based notifications.

lincoln-lm commented on discordbot/OverworldDiscordBot.py:
```diff
@@ -0,0 +1,282 @@
+
+    async def send_discord_event(self, embed, file, channel_id):
+        channel = self.get_channel(int(channel_id))
+        asyncio.run_coroutine_threadsafe(channel.send(embed=embed,file=file),self.loop)
```
asyncio.run_corotine_threadsafe does not need to be used in an async function, await is preferred

lincoln-lm commented on discordbot/OverworldDiscordBot.py:
```diff
@@ -0,0 +1,282 @@
+                self.thread = Thread(target=self.reader_func,args=(ctx,))
+                self.thread.start()
+            except socket.timeout:
+                _message = "Unable to connect to IP {}. Check that the Switch is available.".format(self.config["IP"])
```
using ``f"Unable to connect to IP {self.config["IP"]}...`` is preferred over .format  (this goes for all strings)

jsb-pokemon replied:
Removed all uses of .format().

lincoln-lm commented on discordbot/OverworldDiscordBot.py:
```diff
@@ -0,0 +1,282 @@
+
+import asyncio
+import io
+import json
```
unused import as config is passed as a json already

lincoln-lm commented on discordbot/OverworldDiscordBot.py:
```diff
@@ -0,0 +1,282 @@
+import json
+import socket
+import sys
+from functools import lru_cache
+from threading import Thread
+from time import sleep
+from enum import Enum
+from distutils.util import strtobool
```
imports should be sorted by type and then alphabetically for cleanliness

lincoln-lm commented on discordbot/OverworldDiscordBot.py:
```diff
@@ -0,0 +1,282 @@
+    return pkm.brilliant
+
+def is_pkm_marked(pkm):
+    return pkm.mark > 52 and pkm.mark < 98 and pkm.mark != 69
```
\> 52 and < 98 checks aren't necessarily needed as an overworld PK8 will not spawn with a ribbon, ``pkm.mark != 255`` is simpler

lincoln-lm commented on discordbot/OverworldDiscordBot.py:
```diff
@@ -0,0 +1,282 @@
+
+def get_pkm_mark_string(pkm):
+    return 'No Mark' if pkm.mark == 255 else pkm.Ribbons[pkm.mark]
+
+# function that formats the pkm into a title string and a description string for discord embeds
+def pkm_format(pkm, context):
+    # get strings from pkm to use
+    # 1/2/H
+    ability_string = str(pkm.ability) if pkm.ability < 4 else 'H'
+    # Brilliant/Not Brilliant
+    brilliant_string = 'Brilliant' if pkm.brilliant else 'Not Brilliant'
+    # ♂/♀/-
+    gender_emoji = Util.GenderSymbol[pkm.gender-1]
+    # X/X/X/X/X/X
+    ivs_string = '/'.join(str(x) for x in pkm.ivs)
+    # pulls an emoji from the current server with the id of the mark as its name
+    # will throw an error unless you have all the marks as emojis
+    mark_emoji = (str(get(context.message.guild.emojis, name=str(pkm.mark))) + ' ') if pkm.mark != 255 else ''
+    # No Mark/Mark Name
+    mark_string = get_pkm_mark_string(pkm)
+    # Nature string (Jolly,Adamant,Timid,etc.)
+    nature_string = Util.STRINGS.natures[pkm.nature]
+    #  /⋆/◇
+    shiny_string = '' if pkm.shinyType == 0 else '⋆ ' if pkm.shinyType == 1 else '◇ '
+    # Species-Form Number (Articuno-1 for GCuno)
+    species_string = get_pkm_species_string(pkm)
+
+    # format of the title
+    title = f"{mark_emoji}{gender_emoji} {shiny_string} {species_string}"
+    # format of the description
+    description = f"{nature_string} {ability_string}\n{ivs_string}\n{mark_string}\n{brilliant_string}"
+
+    return title, description
+
+# function to pull pokemon images from pokebase, has a cache since the same pokemon tend to spawn in an area
+@lru_cache(maxsize=32)
+def get_pokemon(species,shiny):
+    image_bytes = pb.SpriteResource('pokemon', species, shiny=shiny).img_data
+    im = Image.open(io.BytesIO(image_bytes)).convert('RGBA')
+    return im
```
non class functions defined in a class file, preferably would be added as ``@staticmethod`` in some way, or made into their own class. Note for future: a class and command dedicated to filtering would be nice

jsb-pokemon replied:
Agreed. Adding this as tech debt to the readme.

lincoln-lm commented on discordbot/OverworldDiscordBot.py:
```diff
@@ -0,0 +1,282 @@
+from nxreader import SWSHReader
+# from . import Channels
+
+class Channels(Enum):
+    NoChannel = 0
+    EventNotificationChannelIdForShiny = 1
+    EventNotificationChannelIdForBrilliant = 2
+    EventNotificationChannelIdForMark = 3
+    EventNotificationChannelIdForRareMark = 4
+    NotificationChannelForInfo = 5
+    TextNotificationChannelIdForShiny = 6
+    TextNotificationChannelIdForBrilliant = 7
+    TextNotificationChannelIdForMark = 8
+    TextNotificationChannelIdForRareMark = 9
```
Note for future: to go with the filtering class below, channels would also need to be reworked to fit the custom filters

jsb-pokemon replied:
Added as a future improvement in the readme.

lincoln-lm commented on discordbot/OverworldDiscordBot.py:
```diff
@@ -0,0 +1,282 @@
+# Needs a discord bot account and a server with all the mark ids as emojis.
```
as it is a class in the reader now, comments like this should be moved to the readme along with a full section on the discord bot, section should include/link to a guide on how to set up a bot and obtain its token

lincoln-lm replied:
Note for future: command/option to have the bot automatically add mark emojis to your discord server, and to have them stored for use with the gui instead of pulling from serebii

jsb-pokemon replied:
Added a tiny section to the readme, removed the comment, and added new items to the future improvements list.

lincoln-lm commented on discordbot/OverworldDiscordBot.py:
```diff
@@ -0,0 +1,282 @@
+    def __init__(self):
+        # config to refactor out later
+        # prefix for bot commands
+        bot_prefix = "$"
```
having bot_prefix be in config for more customization is nice

jsb-pokemon replied:
This prefix gets defined and passed in during construction, but config doesn't get passed in until afterwards (init even sets self.config to a default empty value). How do we set the bot_prefix from the config.json?

lincoln-lm replied:
none of the discord bot commands are used/usable before the config is passed correct? ``super().__init__(command_prefix=bot_prefix)`` could be delayed to only run after the config is passed

jsb-pokemon replied:
Done and done. I also simplified configuration while fixing this.

lincoln-lm commented on discordbot/OverworldDiscordBot.py:
```diff
@@ -0,0 +1,282 @@
+    def reader_func(self,ctx):
+        # config to refactor out later
+        # color for the line on the side of the embed, 0xfad1ff is pink
+        embed_color = 0xfad1ff
```
having embed_color be in config for more customization is nice, along with string equivalents for those who dont want to use hex codes (red,green,purple,etc.)

jsb-pokemon replied:
Added as a future improvement in the readme.

lincoln-lm commented on discordbot/OverworldDiscordBot.py:
```diff
@@ -0,0 +1,282 @@
+    return 'No Mark' if pkm.mark == 255 else pkm.Ribbons[pkm.mark]
+
+# function that formats the pkm into a title string and a description string for discord embeds
+def pkm_format(pkm, context):
+    # get strings from pkm to use
+    # 1/2/H
+    ability_string = str(pkm.ability) if pkm.ability < 4 else 'H'
+    # Brilliant/Not Brilliant
+    brilliant_string = 'Brilliant' if pkm.brilliant else 'Not Brilliant'
+    # ♂/♀/-
+    gender_emoji = Util.GenderSymbol[pkm.gender-1]
+    # X/X/X/X/X/X
+    ivs_string = '/'.join(str(x) for x in pkm.ivs)
+    # pulls an emoji from the current server with the id of the mark as its name
+    # will throw an error unless you have all the marks as emojis
+    mark_emoji = (str(get(context.message.guild.emojis, name=str(pkm.mark))) + ' ') if pkm.mark != 255 else ''
+    # No Mark/Mark Name
+    mark_string = get_pkm_mark_string(pkm)
+    # Nature string (Jolly,Adamant,Timid,etc.)
+    nature_string = Util.STRINGS.natures[pkm.nature]
+    #  /⋆/◇
+    shiny_string = '' if pkm.shinyType == 0 else '⋆ ' if pkm.shinyType == 1 else '◇ '
+    # Species-Form Number (Articuno-1 for GCuno)
+    species_string = get_pkm_species_string(pkm)
+
+    # format of the title
+    title = f"{mark_emoji}{gender_emoji} {shiny_string} {species_string}"
+    # format of the description
+    description = f"{nature_string} {ability_string}\n{ivs_string}\n{mark_string}\n{brilliant_string}"
+
+    return title, description
```
Note for future: have a way to make formatting easier for the user, assume little knowledge of python

jsb-pokemon replied:
Unsure if this is needed since the cards are a one-time design.

lincoln-lm replied:
i would say it wouldn't be high priority, but some users might wish to have their embeds look different and it would be a nice option

jsb-pokemon replied:
Added to the readme!

lincoln-lm commented on tests/OverworldDiscordBot.py:
```diff
@@ -0,0 +1,22 @@
+# Needs a discord bot account and a server with all the mark ids as emojis
+
+import asyncio
+import io
+import json
+import sys
+from functools import lru_cache
+from threading import Thread
+from time import sleep
+
+import discord
+import pokebase as pb
+from discord.ext import commands
+from discord.utils import get
+from PIL import Image
```
unused imports, only ``import json`` and ``import sys`` are needed

jsb-pokemon replied:
Fixed.

lincoln-lm commented on tests/OverworldDiscordBot.py:
```diff
@@ -0,0 +1,22 @@
+
+bot = OverworldDiscordBot()
+bot.configure(json.load(open("../config.json")))
+bot.run("ADD-TOKEN-HERE")
```
having token be in config/a secure config to avoid needing to actually edit script is helpful, another temporary solution could be using an ``input()``?

jsb-pokemon replied:
Is there a convention for something like config.json being excluded in a gitignore file, and the config.json is stored in something like a config.json.template file? So the user basically copies it, fills in config, and then it never gets checked in?

lincoln-lm replied:
including config.json as config.json.template and then gitignoring config.json would work, the user would have to copy config.json.template, remove .template, and then edit it for it to be used though. an instruction to do this would need to be added to readme

jsb-pokemon replied:
Added some instructions to the readme, though they'll change before too long as this becomes even more usable.

lincoln-lm commented on config.json:
```diff
@@ -1,3 +1,13 @@
 {
-    "IP": "192.168.0.106"
+    "IP": "",
+    "EventNotificationChannelIdForShiny": "",
+    "EventNotificationChannelIdForBrilliant": "",
+    "EventNotificationChannelIdForMark": "",
+    "EventNotificationChannelIdForRareMark": "",
+    "NotificationChannelForInfo": "",
+    "TextNotificationChannelIdForShiny": "",
+    "TextNotificationChannelIdForBrilliant": "",
+    "TextNotificationChannelIdForMark": "",
+    "TextNotificationChannelIdForRareMark": "",
+    "EnableDebugLogging": "False"
```
possibly add token here, and commands to change and save the config on the fly would avoid needing to edit more files

jsb-pokemon replied:
Token added and test updated to accept it, so all config is now in config.json. I'll create a "Possible Future Improvements" section at the end of README.md and start recording things like "update config from discord on the fly" to it.

lincoln-lm commented on discordbot/OverworldDiscordBot.py:
```diff
@@ -0,0 +1,282 @@
+            _message = "Overworld Discord Bot has started."
+            await self.send_discord_msg(_message, Channels.NotificationChannelForInfo)
+            # change bot presence to "Watching some ram for shinies"
+            await self.change_presence(activity=discord.Activity(type=discord.ActivityType.watching, name="some ram for shinies"))
```
Note for future: customizable activities in the form of config/command

jsb-pokemon replied:
What kind of activities are you thinking? Can you give an example?

lincoln-lm replied:
> What kind of activities are you thinking? Can you give an example?

discord bots can have playing/streaming/listening/watching activities with custom text after them, they cant have full custom statuses but customization to what their activity says is always nice

one use i could think of is an auto updating activity to show how many total overworld encounters have been read

jsb-pokemon replied:
Adding to the feature list on the readme!

lincoln-lm commented on discordbot/OverworldDiscordBot.py:
```diff
@@ -0,0 +1,282 @@
+                _message = "Unable to connect to IP {}. Check that the Switch is available.".format(self.config["IP"])
+                await self.send_discord_msg(_message, Channels.NotificationChannelForInfo)
+
+        # function to run whenever the ping command is called
+        @self.command()
+        async def ping(ctx):
+            # say the bots latency in discord channel
+            await ctx.send(f'Pong! {round(self.latency, 3)}')
```
Note for future: use MS instead of self.latency

jsb-pokemon replied:
Can we do this here and now? What would this look like?

lincoln-lm replied:
this can be done by either doing something like seen [here](https://github.com/poketwo/poketwo/blob/8b0ba184dfde87298d63602cd23982491a7f4737/cogs/bot.py#L339), or it can be done by converting the number given  by [.latency](https://discordpy.readthedocs.io/en/stable/api.html?highlight=latency#discord.Client.latency) to milliseconds

jsb-pokemon replied:
I'm going to skip this for now. If you want to do this go ahead!

lincoln-lm commented on discordbot/OverworldDiscordBot.py:
```diff
@@ -0,0 +1,282 @@
+            await self.send_discord_msg(_message, Channels.NotificationChannelForInfo)
+            # change bot presence to "Watching some ram for shinies"
+            await self.change_presence(activity=discord.Activity(type=discord.ActivityType.watching, name="some ram for shinies"))
+
```
Note for future: save stats and have a command to display them (time running, overworld mons read, shinies. how many of each mark(?), etc.

jsb-pokemon replied:
Added as a future improvement in the readme.

lincoln-lm commented on discordbot/OverworldDiscordBot.py:
```diff
@@ -0,0 +1,282 @@
+        channel = self.get_channel(int(channel_id))
+        asyncio.run_coroutine_threadsafe(channel.send(embed=embed,file=file),self.loop)
+
+    async def send_discord_msg(self, message, destination=Channels.NoChannel):
```
a parameter and filter for whether or not to ping, along with the id of the user/role to ping added to config would be helpful for if you want to log some things but only ping for others. Possibly needs to wait for the filter class in the future

jsb-pokemon replied:
Added as a future improvement in the readme.

lincoln-lm commented on discordbot/OverworldDiscordBot.py:
```diff
@@ -0,0 +1,282 @@
+                # for each pkm check for filter
+                for pkm in pkms:
+                    # Determine if the pokemon info should be an event notification, and where it routes to. Send text notifications inline.
+                    _channels_to_notify = []
```
do these variables need to be marked as internal/private when they are only accessible from inside the thread and have no use outside of it?

jsb-pokemon replied:
I like the convention of differentiating from class variables that are members and local throwaways. Is there a common python convention for this?

lincoln-lm replied:
using _ itself as a variable indicates something to throwaway that isnt used at all, using it to prepend variables is typically used to mark a variable as private, or something that should not be accessed outside of the class (``self._secret_variable = 123``)

lincoln-lm replied:
having variables named like this is fine, but it's better to be consistent across the project and i havent used that syntax elsewhere

jsb-pokemon replied:
OK I can remove that for consistency.

lincoln-lm reviewed (changes requested):
Overall, has a lot of good new additions! A few tweaks needed for cleanliness and a lot of comments for possibilities, the notes for future are things i can handle/work with you on if you don't decide to/don't have the time to do yourself.

Lincoln-LM commented:
oh and as a last note, commiting bigger changes separately is preferred as opposed to singular big commits for version history in the future

jsb-pokemon committed:
First pass of tweaking changes based on PR feedback.

jsb-pokemon committed:
More PR feedback changes.

lincoln-lm commented on README.md:
```diff
@@ -46,6 +46,11 @@
 * "GUIOverworld.py" Is the script you need to run if you want GUI reading of Overworld Pokémon
 * The other scripts are for dumping info and can be mostly ignored for the average person
 
+### Usage: sOverworld Scanner Discord Bot
```
is "sOverworld" intentional?

jsb-pokemon replied:
Nope! Haha, fixed.

lincoln-lm commented on config.json:
```diff
@@ -1,13 +1,14 @@
     "TextNotificationChannelIdForMark": "",
-    "TextNotificationChannelIdForRareMark": "",
-    "EnableDebugLogging": "False"
+    "TextNotificationChannelIdForRareMark": "883789388289568818"
```
remember to set config back to default, also should regenerate your bots token so no one can use it

jsb-pokemon replied:
Yeeep, discord caught it and disabled it. I went ahead and did the template thing and excluded config.json using gitignore to prevent this from happening again.

jsb-pokemon committed:
Accidentally committed local config. Removed this, deleting config.json, and adding a template file instead. config.json is now ignored and can't be accidentally committed.

jsb-pokemon committed:
Removed leading underscore from variables for consistency.

jsb-pokemon committed:
Updating the readme to reflect more changes and ideas.

jsb-pokemon committed:
Removed all use of .format() on strings.

jsb-pokemon committed:
Simplied configuration for the discord bot by requiring it in the constructor.

DiscordBotPrefix, used to prefix commands to the bot sent from discord, is now part of the config.

lincoln-lm commented on README.md:
```diff
@@ -46,10 +46,14 @@
 
-The bot can be run once *config.json* is filled in, by running *tests/OverworldDiscordBot.py*
+In order to run the bot do the following:
+1. Copy *config.template.json* to *config.json*
+1. Fill out *config.json* with appropriate values.
+1. Execute *tests/OverworldDiscordBot.py*
+1. From the discord server, run the *$start* command to start the bot's main scanner thread.
```
steps are all the first step

Lincoln-LM commented:
Everything here looks good so I'm going to go ahead and accept the PR and correct the readme myself
